<a href="https://colab.research.google.com/github/codyub/ESAA/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%99%84%EB%B2%BD_%EA%B0%80%EC%9D%B4%EB%93%9C_ch8_1~3_pg_466~486.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ch8 1~3 텍스트분석

# 01 텍스트 분석 이해

- 텍스트 분석은 비정형 데이터 분석
- ML 알고리즘은 숫자형 피처 기반 데이터만 입력받을 수 있으므로, 텍스트를 머신러닝에 적용하기 위해서는 비정형 텍스트 데이터를 어떻게 피처 형태로 추출하고 의미 있는 값을 부여하는가가 중요
- BOW, Word2Vec 등이 대표적인 텍스트 피처 벡터화

# 텍스트 분석 수행 프로세스
1. 텍스트 사전 준비작업(텍스트 전처리) : 텍스트를 피처로 만들기 전에 미리 클렌징, 대/소문자 변경, 특수문자 삭제 등의 클렌징 작업, 단어(word) 등의 토큰화 작업, 의미 없는 단어(stop word) 제거 작업, 어근 추출(stemming, lemmatization) 등의 텍스트 정규화 작업.
2. 피처 벡터화/추출 : 가공된 텍스트에서 피처룰 추출하고 여기에 벡터값 할당. BOW의 Count 기반과 TF-IDF 기반 벡터화.
3. ML 모델 수립 및 학습/예측/평가: ML 모델 적용

# 파이썬 기반의 NLP, 텍스트 분석 패키지
- NLTK(Natural Language Toolkit for Python) : 방대한 데이터 세트와 서브 모듈을 갖고 있음. NLP의 거의 모든 영역 커버. 하지만 수행 속도 측면에서 아쉬운 부분이 있어 대량의 데이터 기반에서는 제대로 활용 못하고 있다.
- Gensim : 토픽 모델링에서 두각. Word2Vec 등의 다양한 신기능도 제공.
- SpaCy : 뛰어난 수행 성능으로 최근 가장 주목.

- 사이킷런은 머신러닝 위주의 라이브러리라서 NLP 특화 라이브러리(ex. 어근처리)는 없다. 
- 하지만 텍스트를 일정 수준으로 가공하고 머신러닝 알고리즘에 텍스트 데이터를 피처로 처리하기 위한 편리한 기능 제공해서 충분히 텍스트 분석 가능.

# 02 텍스트 사전 준비 작업(텍스트 전처리)-텍스트 정규화
- 텍스트 자체를 바로 피처로 만들 수는 없으므로, 클렌징, 토큰화, 필터링/스톱워드제거/철자 수정, stemming, lemmatization 과정이 필요함



# 클렌징
- 텍스트 분석에 오히려 방해가 되는 불필요한 문자, 기호 등을 사전에 제거하는 작업

# 텍스트 토큰화
- 문장 토큰화: 문서에서 문장을 분리
- 단어 토큰화: 문장에서 단어를 토큰으로 분리

# 문장 토큰화
- 문장의 마침표, 개행문자 등 문장의 마지막을 뜻하는 기호에 따라 분리하는 것이 일반적

- 3개의 문장으로 이뤄진 텍스트 문서를 문장으로 각각 분리하는 예제

In [1]:
from nltk import sent_tokenize
import nltk
nltk.download('punkt') #마침표, 개행 문자 등의 데이터 세트 다운로드

text_sample='The Matrix is everywhere its all around us, here even in this room. \nYou can see it out your window or on your television. \nYou feel it when you go to work, or go to church or pay your taxes.'
sentences=sent_tokenize(text=text_sample)
print(type(sentences), len(sentences))
print(sentences)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


<class 'list'> 3
['The Matrix is everywhere its all around us, here even in this room.', 'You can see it out your window or on your television.', 'You feel it when you go to work, or go to church or pay your taxes.']


- 반환하는 것은 각 문장으로 구성된 list 객체
- 반환된 list 객체가 3개의 문장으로 된 문자열을 갖고 있는 것을 알 수 있음

# 단어 토큰화
- 문장을 단어로 토큰화, 기본적으로 공백, 콤마, 마침표, 개행문자 등으로 단어를 분리, 정규표현식을 이용해 다양한 유형으로 토큰화를 수행 가능
- BOW와 같이 순서가 중요하지 않은 경우, 문장 토큰화를 사용하지 않고 단어 토큰화만 사용해도 됨
- 문장토큰화는 의미가 중요한 요소로 사용될 때 사용

In [2]:
from nltk import word_tokenize

sentence='The Matrix is everywhere its all around us, here even in this room.'
words=word_tokenize(sentence)
print(type(words), len(words))
print(words)

<class 'list'> 15
['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.']


- sent_tokenize와 word_tokenize 조합해 문서에 대해 모든 단어를 토큰화
>- 문서 샘플을 먼저 문장으로 나누고 개별 문장을 다시 단어로 토큰화하는 함수를 생성

In [3]:
from nltk import word_tokenize, sent_tokenize

#여러 개의 문장으로 된 입력 데이터를 문장별로 단어 토큰화하게 만드는 함수 생성
def tokenize_text(text):
  #문장별로 분리 토큰
  sentences=sent_tokenize(text)
  #분리된 문장별 단어 토큰화
  word_tokens=[word_tokenize(sentence) for sentence in sentences]
  return word_tokens

#여러 문장에 대해 문장별 단어 토큰화 수행
word_tokens=tokenize_text(text_sample)
print(type(word_tokens), len(word_tokens))
print(word_tokens)

<class 'list'> 3
[['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.'], ['You', 'can', 'see', 'it', 'out', 'your', 'window', 'or', 'on', 'your', 'television', '.'], ['You', 'feel', 'it', 'when', 'you', 'go', 'to', 'work', ',', 'or', 'go', 'to', 'church', 'or', 'pay', 'your', 'taxes', '.']]


- 3개 문장을 문장별로 먼저 토큰화 했으므로, word_tokens 변수는 3개의 리스트 객체를 내포하는 리스트
- 내포된 개별 리스트 객체는 각각 문장별로 토큰화된 단어를 요소로 갖고 있음
- 문장을 단어별로 하나씩 토큰화 할 경우 문맥적인 의미는 무시 -> 이러한 문제를 조금이라도 해결해 보고자 도입된 것이 n-gram
- n-gram은 연속된 n개의 단어를 하나의 토큰화 단위로 분리해 내는 것
-> n개 단어 크기 윈도우를 만들어 문장의 처음부터 오른쪽으로 움직이면서 토큰화를 수행

# 스톱 워드 제거
- 스톱 워드: 분석에 큰 의미가 없는 단어. ex. 문법 요소 등. 빈번해서 오히려 중요 단어로 인지될 수 있어, 제거하는 것이 중요.
- 언어별로 스톱 워드가 목록화 되어 있음. 
- NTLK의 스톱 워드에는 어떤 것이 있을지 확인.

- NLTK의 스톱워드 목록 내려 받기

In [4]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [5]:
print('영어 stop word 개수:', len(nltk.corpus.stopwords.words('english')))
print(nltk.corpus.stopwords.words('english')[:20])

영어 stop word 개수: 179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his']


- 영어의 경우 스톱 워드가 179개.

- 스톱워드를 필터링으로 제거해 분석을 위한 의미 있는 단어만 추출

In [6]:
import nltk

stopwords=nltk.corpus.stopwords.words('english')
all_tokens=[]
#위 예제에서 3개의 문장별로 얻은 word_tokens list에 대해 스톱 워드를 제거하는 반복문
for sentence in word_tokens:
  filtered_words=[]
  #개별 문장별로 토큰화된 문장 list에 대해 스톱 워드를 제거하는 반복문
  for word in sentence:
    #소문자로 모두 변환
    word=word.lower()
    #토큰화된 개별 단어가 스톱 워드의 단어에 포함되지 않으면 word_tokens에 추가
    if word not in stopwords:
      filtered_words.append(word)
    all_tokens.append(filtered_words)

print(all_tokens)

[['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['matrix', 'everywhere', 'around', 'us', ',', '

- in, this 등이 필터링을 통해 제거됐음을 알 수 있음

# Stemming과 Lemmatization
- 문법적인 요소에 따라 단어가 다양하게 변하는 언어들이 많은데, 이는 문법적 또는 의미적으로 변화하는 단어의 원형을 찾는 것
- Lemmatization: 정교하며, 의미론적인 기반에서 단어의 원형을 찾음, 품사와 같은 문법적인 요소와 더 의미적인 부분을 감안해 정확한 처랒로 된 어근 단어를 찾아줌. 더 오랜 시간의 변환이 필요함
- Stemming: 우너형 단어로 변환 시 일반적인 방법을 적용하거나 더 단순화된 방법을 적용해 원래 단어에서 일부 철자가 훼손된 어근 단어를 추출하는 경향이 있음
- NLTK는 다양한 stemmer를 제공. porter, lancaster, snowball stemmer. lemmatization을 위해서는 wordnetlemmatization제공. 

- stemmer와 lemmatization 비교

In [7]:
#stemming
from nltk.stem import LancasterStemmer
stemmer=LancasterStemmer()

print(stemmer.stem('working'), stemmer.stem('works'), stemmer.stem('worked'))
print(stemmer.stem('amusing'), stemmer.stem('amuses'), stemmer.stem('amused'))
print(stemmer.stem('happier'), stemmer.stem('happiest'))
print(stemmer.stem('fancier'), stemmer.stem('fanciest'))

work work work
amus amus amus
happy happiest
fant fanciest


- 원형을 잘 찾지 못하는 상황 발생

In [8]:
#lemmatization
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')

lemma=WordNetLemmatizer()
print(lemma.lemmatize('amusing', 'v'), lemma.lemmatize('amuses', 'v'), lemma.lemmatize('amused', 'v'))
print(lemma.lemmatize('happier', 'a'), lemma.lemmatize('happiest', 'a'))
print(lemma.lemmatize('fancier', 'a'), lemma.lemmatize('fanciest', 'a'))

[nltk_data] Downloading package wordnet to /root/nltk_data...


amuse amuse amuse
happy happy
fancy fancy


- lemmatization이 stemming보다 더 정확.

# 03 Bag of Words - BOW
- BOW 모델은 문서가 가지는 모든 단어를 문맥이나 순서를 무시하고 일괄적으로 단어에 대해 빈도 값을 부여해 피처 값을 추출하는 모델
-
1. 문장1과 문장2에 있는 모든 단어에서 중복 제거, 각 단어를 칼럼 형태로 나열. 그리고 각 단어에 고유의 인덱스를 부여.
2. 개별 문장에서 해당 단어가 나타나는 횟수를 각 단어(인덱스)에 기재.

# 장점
- 쉽고 빠른 구축
# 단점
- 문맥 의미 반영 부족 : 단어의 순서를 고려하지 않기 때문에, 문장 내에서 단어의 문맥적인 의미가 무시. 보완하기 위해 n-gram 사용할 수 있지만 제한적임.
- 희소 행렬 문제(희소성, 희소 생렬) : BOW로 피처 벡터화를 수행하면 희소 행렬 형태의 데이터 세트가 만들어지기 쉬움. 많은 단어가 칼럼으로 만들어지지만, 단어가 문서마다 나타나지 않는 경우가 훨씬 많음. 대부분의 데이터가 0으로 채워지게 된다. 이처럼 대규모의 칼럼으로 구성된 행렬에서 대부분의 값이 0으로 채워지는 행렬을 희소 행렬이라고 한다. <-> 밀집 행렬. 희소 행렬은 머신러닝 알고리즘의 수행 시간과 예측 성능을 떨어뜨려 특별한 기법이 필요.

# BOW 피처 벡터화
- 텍스트는 특정 의미를 가지는 숫자형 값인 벡터 값으로 변환해야 하는데, 이러한 변환을 피처 벡터화라고 함
- 각 문서이 텍스트를 단어로 추출해 피처로 할당하고, 각 단어의 발생 빈도와 같은 값을 이 피처에 값으로 부여해 각 문서를 이 단어 피처의 발생 빈도 값으로 구성된 벡터로 만드는 기법
- 피처 벡터화는 기존 텍스트 데이터를 또 다른 형태의 피처의 조합으로 변경하기 때문에 넓은 범위의 피처 추출에 포함
- BOW 모델에서 피처 벡터화를 수행한다는 것은, 모든 문서에서 모든 단어를 칼럼 형태로 나열하고 각 문서에서 해당 단어의 횟수나 정규화된 빈도를 값으로 부여한느 데이터 세트 모델로 변경하는 것

# BOW의 피처 벡터화 종류
1. 카운트 기반의 벡터화
- 단어 피처에 값을 부여할 때, 각 문서에서 해당 단어가 나타나는 횟수를 부여하는 경우
- 카운트 값이 높을수록 중요한 단어로 인식
- 카운트만 부여할 경우, 그 문서의 특징을 나타내기 보다 언어의 특성상 문장에서 자주 사용될 수밖에 없는 단어까지 높은 값을 부여하게 된다.
2. TF-IDF 기반의 벡터화
- 개별 문서에서 자주 나타나는 단어에 높은 가중치를 주되, 모든 문서에서 전반적으로 자주 나타나는 단어에 대해서는 패널티를 주는 방식으로 값을 부여
- 몯ㄴ 문서에 반복적으로 자주 발생하는 단어에 대해서는 패널티를 부여하는 방식으로 단어에 대한 가중치의 균형을 맞춰줌


# 사이킷런의 count 및 TF-IDF 벡터화 구현: CountVectorizer, TfidfVectorizer

- CountVectorizer: 카운트 기반의 벡터화를 구현한 클래스, 단지 피처 벡터화만 수행하는게 아니라 소문자 일괄 변환, 토큰화, 스톱 워드 필터링 등의 전처리도 수행해준다. 입력 파라미터 설정 필요.

- max_df : 전체 문서에서 너무 높은 빈도수 가지는 단어 피처를 제외.
- min_df : 전체 문서에서 너무 낮은 빈도수 가지는 단어 피처를 제외.
- max_features : 추출하는 피처의 개수를 제한.
- stop_words : 스톱 워드로 지정된 단어는 추출에서 제외.
- n_gram_range : 단어 순서를 어느 정도 보강하기 위한 n_gram 범위 설정.
- analyzer : 피처 추출을 수행한 단위를 지정. 디폴트는 당연히 word.
- token_pattern : 토큰화를 수행하는 정규 표현식 패턴 지정.
- tokenizer : 토큰화를 별도의 커스텀 함수로 이용시 적용.

보통 사이킷런의 CountVectorizer 클래스를 이용해 카운트 기반의 피처 여러 개의 문서로 구성된 텍스트의 피처 벡터화 방법은
1. 영어의 경우 소문자 변경 등 사전 데이터 가공
2. 디폴트로 단어 기준으로 n_gram_range 반영해 각 단어 토큰화
3. 텍스트 정규화 수행. 어근 변환은 CountVectorizer에서 직접 지원하지 않고 tockenizer 파라미터에 커스텀 어근 변환 함수를 적용하여 수행할 수 있다.
4. 피처 벡터화. 다른 파라미터를 이용해 토큰화된 단어를 피처로 추출하고 단어 빈도수 벡터 값을 적용


- TF-IDF 벡터화는 TfidfVectorizer 클래스 이용. 파라미터 변환 방법은 count와 동일.

# BOW 벡터화를 위한 희소 행렬
- 사이킷런의 countvectorizer, tfidfvectorizer를 이용해 텍스틀 피처 단위로 벡터화해 변환하고 CSR 형태의 희소 행렬을 반환
- 희소행렬: 모든 문서에 있는 단어를 중복을 제외하고 피처로 만들면 많은 피처 칼럼이 만들어짐 ->  하지만 대규모의 행렬에서 레코드의 각 문서가 가지는 단어의 수는 제한적이라 이 행렬의 값은 대부분 0이 차지하고, 이러한 행렬을 의미
- BOW 형태를 가진 언어 모델의 피처 벡터화는 대부분 희소 행렬. 
- 불필요한 0 때문에 메모리 공간이 많이 필요하며, 행렬의 크기가 커서 연산 시에도 시간이 많이 소모됨 -> 따라서 희소 행렬을 변환해야 함: 대표적으로 COO 형식과 CSR 형식. 일반적으로 더 뛰어난 CSR 사용.

# 희소 행렬 - COO 형식
- 0이 아닌 데이터만 별도의 데이터 배열에 저장하고, 그 데이터가 가리키는 행과 열의 위치를 별도의 배열로 저장하는 방식
- 희소 행렬 변환을 위해서는 주로 사이파이를 사용함. 

- 먼저 [ [3,0,1], [0,2,0] ]을 넘파이의 객체로 만들어 COO 형식의 희소 행렬로 변환

In [9]:
import numpy as np

dense=np.array([[3,0,1], [0,2,0]])

- 0이 아닌 데이터를 별도의 배열 데이터로 만들고, 행/열 위치 배열을 각각 만든 후 coo_matirx() 내에 생성 파라미터로 입력

In [10]:
from scipy import sparse

#0이 아닌 데이터 추출
data=np.array([3,1,2])

#행 위치와 열 위치를 각각 배열로 생성
row_pos=np.array([0,0,1])
col_pos=np.array([0,2,1])

#sparse 패키지의 coo_matirx를 이용해 COO 형식의 희소 행렬 생성
sparse_coo=sparse.coo_matrix((data, (row_pos, col_pos)))

- toarray() 메서드로 다시 밀집 형태의 행렬로 출력

In [11]:
sparse_coo.toarray()

array([[3, 0, 1],
       [0, 2, 0]])

# 희소 행렬 - CSR 형식
- COO 형식이 행과 열의 위치를 나타내기 위해 반복적인 위치 데이터를 사용해야 하는 문제점을 해결한 방식
- 행 위치 배열 내에 있는 고유한 값의 시작 위치만 다시 별도의 위치 배열로 가지는 변환 방식

In [12]:
from scipy import sparse

dense2=np.array([[0,0,1,0,0,5],
                 [1,4,0,3,2,5],
                 [0,6,0,3,0,0],
                 [2,0,0,0,0,0],
                 [0,0,0,7,0,8],
                 [1,0,0,0,0,0]])

#0이 아닌 데이터 추출
data2=np.array([1,5,1,4,3,2,5,6,3,2,7,8,1])

#행 위치와 열 위치를 각각 array로 생성
row_pos=np.array([0,0,1,1,1,1,1,2,2,3,4,4,5])
col_pos=np.array([2,5,0,1,3,4,5,1,3,0,3,5,0])

#COO 형식으로 변환
sparse_coo=sparse.coo_matrix((data2, (row_pos, col_pos)))

#행 위치 배열의 고유한 값의 시작 위치 인덱스를 배열로 생성 + 맨 끝에는 총 데이터 수
row_pos_ind=np.array([0,2,7,9,10,12,13])

#CSR 형식으로 변환
sparse_csr=sparse.csr_matrix((data2, col_pos, row_pos_ind))

print('COO 변환된 데이터가 제대로 되었는지 다시 dense로 출력 확인')
print(sparse_coo.toarray())
print('CSR 변환된 데이터가 제대로 되었는지 다시 dense로 출력 확인')
print(sparse_csr.toarray())

COO 변환된 데이터가 제대로 되었는지 다시 dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]
CSR 변환된 데이터가 제대로 되었는지 다시 dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


In [13]:
dense3=np.array([[0,0,1,0,0,5],
                 [1,4,0,3,2,5],
                 [0,6,0,3,0,0],
                 [2,0,0,0,0,0],
                 [0,0,0,7,0,8],
                 [1,0,0,0,0,0]])

coo=sparse.coo_matrix(dense3)
csr=sparse.csr_matrix(dense3)

In [14]:
coo

<6x6 sparse matrix of type '<class 'numpy.int64'>'
	with 13 stored elements in COOrdinate format>

In [15]:
csr

<6x6 sparse matrix of type '<class 'numpy.int64'>'
	with 13 stored elements in Compressed Sparse Row format>